cmd/promtool/backfill.go

```diff
@@ -79,7 +79,16 @@ func createBlocks(input []byte, mint, maxt int64, maxSamplesInAppender int, outp
 
 	var wroteHeader bool
 
+	nextSampleTs := int64(math.MaxInt64)
 	for t := mint; t <= maxt; t = t + blockDuration {
+		if nextSampleTs != math.MaxInt64 && nextSampleTs >= t+blockDuration {
```

Comment:
In L104, you already have `tsUpper := t + blockDuration`. This could be pulled up and you could use `tsUpper` here.